core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -173,26 +174,34 @@ private void refreshContexts() {
         try {
             synchronized (contexts) {
                 final List<Endpoint> newSelectedEndpoints = healthCheckStrategy.getSelectedEndpoints();
+                final Set<Endpoint> newSelectedEndpointsSet = new HashSet<>(newSelectedEndpoints);
```

Comment:
How about checking first that the endpoints are the same? 
```java
final boolean isSameEndpoints = contexts.size() == newSelectedEndpoints.size() &&
                                contexts.keySet().containsAll(newSelectedEndpoints);
```

Reply:
What happens `newSelectedEndpoints` has more endpoints?

Reply:
If the problem happens only when weights are changed, I thought we can use the original logic if they are different. For example:
```java
final boolean isSameEndpoints = contexts.size() == newSelectedEndpoints.size() &&
                                contexts.keySet().containsAll(newSelectedEndpoints);
if (isSameEndpoints) {
    refreshEndpoints();
    return;
}

// same as before
```

Maybe there is something I missed. 🤪

Reply:
If `newSelectedEndpoints` has more endpoints than current endpoints we should add the newly added endpoints.
So we should do this as well:
```java
newSelectedEndpoints.containsAll(contexts.keySet())
```
which iterates another for loop.
Most of the time, the endpoints will be changed so I think it's useless to check if just the weight is changed. 
We will be going to notice that anyway. 😄

Reply:
> newSelectedEndpoints.containsAll(contexts.keySet())

Question: Is this not covered by checking their sizes?

Anyway, this is not a big deal. Let keep it as it is.

Reply:
> Question: Is this not covered by checking their sizes?

Oh, I missed that point. 😅 

> Let keep it as it is.

Yeah, let's keep it as it is.